hal/src/thumbv7em/eic/pin.rs

```diff
@@ -1,11 +1,31 @@
-use crate::gpio::{self, IntoFunction, Port};
+use crate::gpio::{
+    self,
+    v2::FloatingInterrupt,
+    v2::PullUpInterrupt,
+    v2::PullDownInterrupt,
+    Port,
+};
 use crate::target_device;
 
-/// The EicPin trait makes it more ergonomic to convert a gpio pin into an EIC
-/// pin. You should not implement this trait for yourself; only the
-/// implementations in the EIC module make sense.
-pub trait EicPin<T> {
-    fn into_ei(self, port: &mut Port) -> T;
+/// The FloatingEicPin trait makes it more ergonomic to convert a gpio pin into
+/// an EIC pin with a floating input. You should not implement this trait for
+/// yourself; only the implementations in the EIC module make sense.
+pub trait FloatingEicPin<T> {
+    fn into_floating_ei(self, port: &mut Port) -> T;
+}
+
+/// The PullUpEicPin trait makes it more ergonomic to convert a gpio pin into an
+/// EIC pin with a pulled-up input. You should not implement this trait for
+/// yourself; only the implementations in the EIC module make sense.
+pub trait PullUpEicPin<T> {
+    fn into_pull_up_ei(self, port: &mut Port) -> T;
+}
+
+/// The PullDownEicPin trait makes it more ergonomic to convert a gpio pin into
+/// an EIC pin with a pulled-down input. You should not implement this trait for
+/// yourself; only the implementations in the EIC module make sense.
+pub trait PullDownEicPin<T> {
+    fn into_pull_down_ei(self, port: &mut Port) -> T;
 }
```

Comment:
@ianrrees, see [here](https://github.com/ianrrees/atsamd/pull/1) for my proposed implementation of option 2

Reply:
Thanks!  As we've chatted about a little, I do like having `.into_<pull configuration>_ei()` methods, because they make the end user code a bit clearer.  It seems like there should be a way to make the set of three methods for each type without the three traits, but I'm having a hard time seeing it at the moment.

At a slightly higher level, I wonder if we might be better off putting that effort in to a wider-scope EIC rework?

Reply:
If you have the bandwidth for a wider-scope rework, I would say go for it.

I only used external interrupts briefly last year, but I did see all the different configuration options. I'm not sure which of them would make sense to track at compile-time. For the `spi` module, I decided to use types to track anything that could influence the rest of the API at compile-time. For instance, I track the operating mode, because it affects which GPIO pins are required. And I track the transaction length, because it affects the word type. Everything else I left as a run-time defined option.

Reply:
Actually, if you don't have any additional type parameters to add to the `EicPin` struct, you don't really even need to define a new `EicPin` wrapper. You could do it all with traits defined on instances of `Pin<I, Interrupt<C>>`. But if you need to track some other types, you'll need an `EicPin` to do so.

See the [docs](https://atsamd-rs.github.io/docs/samd51n/thumbv7em-none-eabihf/doc/atsamd_hal/typelevel/index.html) for the `typelevel` module for more type-level tricks.

Reply:
Also, here is how you combine the traits. It's more complicated than it needs to be, because the `ExtInt` structs are defined through macros, rather than type-level techniques. It's a bit of a hack, but it works for now.

```rust
pub trait EicPin {
    type Floating;
    type PullUp;
    type PullDown;
    fn into_floating_ei(self, port: &mut Port) -> Self::Floating;
    fn into_pull_up_ei(self, port: &mut Port) -> Self::PullUp;
    fn into_pull_down_ei(self, port: &mut Port) -> Self::PullDown;
}

        impl<MODE: PinMode> EicPin for gpio::$PinType<MODE> {
            type Floating = [<$PadType $num>]<gpio::$PinType<FloatingInterrupt>>;
            type PullUp = [<$PadType $num>]<gpio::$PinType<PullUpInterrupt>>;
            type PullDown = [<$PadType $num>]<gpio::$PinType<PullDownInterrupt>>;
            fn into_floating_ei(self, port: &mut Port) -> Self::Floating {
                [<$PadType $num>]::new(self.into_floating_interrupt(port))
            }
            fn into_pull_up_ei(self, port: &mut Port) -> Self::PullUp {
                [<$PadType $num>]::new(self.into_pull_up_interrupt(port))
            }
            fn into_pull_down_ei(self, port: &mut Port) -> Self::PullDown {
                [<$PadType $num>]::new(self.into_pull_down_interrupt(port))
            }
        }
```

Reply:
Thanks @bradleyharden !  I don't really have the mental bandwidth to re-do the EIC right now, but have applied your suggestion to combine the traits and rebased on to the current master.  My feeling is that this PR is a step toward an ideal solution, but it's certainly not there yet.

As I'm writing this comment, it occurs to me that it might not be easy to get the GPIO pin back from an EIC pin.

Reply:
Good point. Probably just need a `free` method.

Reply:
One other FYI, the way to create an EIC pin for v2::Pin is to use the `new` function directly.